### DISCUSSION TITLE
Maximum pages for static site < 8,000

### DISCUSSION THREAD

**Discussion by coreyward** [vercel/vercel#4009] [General] [answered]
I'm a bit surprised that Vercel has [a hard limit of 16,000 files](https://vercel.com/docs/concepts/limits/overview#files):

> The maximum number of files that can be uploaded when creating a Deployment is 12,500 for source files and 16,000 for build output files.

This is extremely limiting for solutions like Gatsby, where at least two files are generated for every page (both the static HTML and a page-data.json with the data), and many more overall (one for each webpack bundle/chunk). I've worked with a number of 10,000–25,000 page Gatsby sites, and I've had discussions with other developers that use Gatsby with websites that are hundreds of thousands of pages. Is there any plan to change this limit, or should users with larger static sites look to Netlify or other competitors?

**Comment by paulogdm** [comment]
We are currently working on expanding that. However, do notice that for the free-tier the limit may be lower given abuse-protection and other factors.

We should certainly support that number of files easily in a few weeks.

**Reply by coreyward** [reply]
Thanks for the info. Two follow-up questions: 

1. Are the free tier limits on file count published somewhere?
2. Are you also working on increasing the limit on inbound/source files?

**Reply by paulogdm** [reply]
1. Not yet because the free and paid limits are equivalent. However, they will be if we decide to have different limits for different plans.

2. For sure. This is being tested internally already and should be out in the next couple of days/weeks.

**Reply by Aleksion** [reply]
@paulogdm you said a few weeks 104days ago. Is there any update to this? Right now we're forced to migrate to netlify because of this. The built in fallback solution of next seems to be hurting our SEO efforts (google is indexing blank loading pages).

**Comment by qnxdev** [comment]
Is there actually any limits for hosting on vercel?
Any limits that actually matters?
Can my Next.js App deployed on vercel serve content to 100,000 simultaneous users?

**Reply by TheDSCPL** [reply]
It can since it runs on lambdas. The more traffic you have, the more lambdas are spawned to handle it. Just remember to look at the limits and pricing if you plan on exceeding them.

**Reply by coreyward** [reply]
> Can my Next.js App deployed on vercel serve content to 100,000 simultaneous users?

No. You can only have 1,000 concurrently running serverless functions on a Team ([source](https://vercel.com/docs/platform/limits#serverless-function-concurrency)). If each of the 100,000 users only makes a single request every 25 seconds and the serverless function only takes 250ms to execute, you could theoretically support that number of users, but you would not have any additional capacity for serverless functions across your Team.

**Comment by wallawe** [comment]
Has there been an update to this since the question was originally asked? I'm building a site using Next's `getStaticProps` and there will be around 24,000 pages. I'm curious what sort of issues/limitations I should expect to run into.

**Reply by coreyward** [reply]
According to the [documentation](https://vercel.com/docs/platform/limits#files), no. I'd be curious to hear from someone on the Vercel team about this, though.

**Reply by emindeniz99** [reply]
Hi,
Have you done? What happened?

**Comment by Kae7in** [comment]
The current Vercel page limit of 16k only applies to uploaded files and BUILD output files. There is no theoretical limit to pages being built AFTER the build step (during runtime).

You can generate static pages during runtime by setting `fallback: true` property in the getStaticPaths() method. This allows Next.js to incrementally render and store static pages only after they has been requested by a user for the first time. For the first user it will be slow then every request after that will be blazing fast. Vercel has a blog post tackling this exact hypothetical situation where you need to generate 100k static product pages: https://vercel.com/blog/nextjs-server-side-rendering-vs-static-generation

Hope that helps.

**Reply by coreyward** [reply]
> “For the first user it will be slow…”

This is an unacceptable tradeoff. The more painful it is to pre-build pages, the more impactful this effect.

**Reply by Kae7in** [reply]
> > “For the first user it will be slow…”
> 
> This is an unacceptable tradeoff. The more painful it is to pre-build pages, the more impactful this effect.

Would it not be more painful to build thousands of pages at build time? If you have a site with 1M page details and you make even a 1-line change in JS or HTML, the server then has to rebuild all of those pages, which could take a tremendous amount of time. Even if you were hosting on AWS or Google Cloud, I don't think this an approach you'd want to take.

What's your use-case for generating so many pages at build time?

**Reply by coreyward** [reply]
The issues with building on-request and caching are the same as ever: downtime from dependencies, added latency, hardware has to be waiting and ready for the request, etc. A hybrid solution where you build 1M pages on-demand then cache might deploy quickly, but you're going to see negative conversion effects from both users and search engines. If you have to make changes like that several times a day, maybe you shouldn't be relying on what is effectively a Varnish cache for performance and should instead be optimizing server side build times (i.e. a more traditional architecture). 

For what it's worth, there's a huge difference between 16,000 and 1,000,000. In fact, it's 984,000. But we're not talking about page limits, we're talking about *file* limits. Since Gatsby generates more than 2 files per page, you half the number of pages possible (hence the title referencing “<8,000”.  I haven't needed to build 1M pages, but I have worked on projects like an audiobook platform where there is a catalog of around 2,000 audio titles, a few hundred tag pages, a few hundred collection pages, around 1,000 blog posts, a hundred or so blog index pages, a few dozen author pages, and so on. That's around 4,000 pages, generating around 9,000 files (~4k HTML files, ~4k page-data.json files, and a whole bunch of webpack chunks and other pre-built assets). It builds in about 5 minutes without any Gatsby caching magic (a bit longer on Netlify's builds platform since they're somewhat underpowered).

**Reply by joergrech** [reply]
I have a similar problem as @coreyward as we potentially want to pre-render and deploy 5+ million pages with our system. Currently, we only pre-render ~12k pages but would like to pre-render at least 40k. 

I understand that you might not support this in a Pro account but why isn't there a "Pro++" account that could support this? As far as I understand you only had to store the pages in S3, or? Or is this available in the enterprise account?

**Reply by molebox** [reply]
Hey folks 👋

We have updated our [documentation](https://vercel.com/docs/concepts/limits/overview#files) to reflect how best to work with large numbers of output files. We have a maximum of 16,000 output files that can be generated at build time, if you think your application will surpass that number (by the sounds of it, you do) we recommend looking into [Incremental Static Regeneration](https://nextjs.org/docs/basic-features/data-fetching#incremental-static-regeneration) (As Lee mentions in the answer to this discussion), which will allows you to pre-render a subset of the total number of pages at build time.

**Comment by leerob** [comment, marked as answer]
https://vercel.com/docs/concepts/next.js/incremental-static-regeneration

**Reply by coreyward** [reply]
This only works with Next.js though, right? Which would mean it's not possible to use ISR with Gatsby, which is what I referenced in the original question. I don't understand why this is such a hard limit—it's strange that Vercel is so prescriptive about how a customer builds their application. Why not just set it at something reasonably high, like 128,000, and avoid the concern altogether?

**Reply by leerob** [reply]
ISR can work with any framework! Nuxt 3 is going to have support for it 😄 Gatsby has their own version, "Deferred Static Generation". This is enabled through the Vercel File System API: https://vercel.com/docs/file-system-api

It's a hard limit because the majority of users on our platform are not hitting this limit, and those who do, ultimately end up moving to some sort of deferred generation to keep their build times low. Generating 100,000+ static pages isn't a viable solution for most teams.

**Reply by coreyward** [reply]
> ISR can work with any framework! 

Is this documented? All I'm seeing is Next.js-specific documentation.

> It's a hard limit because the majority of users on our platform are not hitting this limit…

I think there's a bit of survivorship bias in that justification. It's because of this hard limit that I have not kicked the tires with Vercel (well, this and the pricing keeps changing and is now really hard to evaluate), and I suspect I'm not alone in this. If you doubled or more all of your hard constraints that are not upstream-enforced (i.e. AWS) and authored a blog post about it, I bet you'd see a sustained bump in growth.

**Reply by leerob** [reply]
Not yet, here's the Nuxt docs: https://v3.nuxtjs.org/
<img width="341" alt="CleanShot 2021-10-14 at 14 18 06@2x" src="https://user-images.githubusercontent.com/9113740/137381697-e1d07c29-bcfb-46e0-8eda-ca7775e6e46a.png">

We'll share more on that soon! It was designed to work with any framework, and we've been working with other frameworks to adopt it 😄

**Comment by Pieparker** [comment]
Do remember that the limit is per _deployment_.

So, you could [make use of rewrites](https://vercel.com/support/articles/how-can-i-serve-multiple-projects-under-a-single-domain?query=rewrite#deploying-rewrite-rules) to "combine" deployments/projects, chaining multiple deployments/projects.
